feat: Provide Process Instance Search Request API in the Zeebe Client

romansmirnov opened the pull request:
## Description

Implements a search request API in the Zeebe Client so users can search process instances creating a query, for example
```java
final SearchQueryResponse<ProcessInstance> response = client.newProcessInstanceQuery()
        .filter((f) -> f.variable((v) -> v.name("foo").eq(5)))
        .sort((s) -> s.endDate().asc())
        .page((p) -> p.size(20))
        .send()
        .join();
```

## Related issues

closes #

tmetzke commented:
I haven't looked at the code yet, will do that tomorrow.

On a logical level, great proposal, @romansmirnov 👍 
The fluent builders are quite handy 👍 
The static builders should come in handy for discovery of options, i.e. what kind of filters and sorts can I create without having to use constructors of classes I first have to find 👍

Regarding command creation and execution, I think it's fine to have only one level of steps, as they are usually called. The options are all equally optional and can be left out completely. That's like a topology command, with no required options to use, so it's directly a final step. That should be okay.

The only thing that looks a bit off is having to use `toArray` on the sort values of the previous response. Would be nice if the returned sort values can be consumed by the query API directly. Not sure if there's a specific reason to have different data types here?

ralfpuchert reviewed (commented):
LGTM, found a minor type. I saw the fluent and builder approach already in demo application code.

nathansandi commented:
I liked the approach over the API, looks pretty fluent in terms of proposal to me. 💯  It effectively covers use cases from simple to complex search requests, demonstrating the versatility of the API.

I would agree with @tmetzke in the terms of use toArray on the sort values of the previous response. But it can be improved after it works.


Great work here 👍

romansmirnov commented:
Updated the API:
1. Removed the usage of `#toArray()` to pass searchAfter/searchBefore values
2. Returns a proper response object

```java
final SearchQueryResponse<ProcessInstance> response = client.newProcessInstanceQuery()
        .filter((f) -> f.variable((v) -> v.name("it").eq(5)))
        .sort((s) -> s.endDate().asc())
        .page((p) -> p.size(20))
        .send()
        .join();

final SearchQueryResponse<ProcessInstance> response2 = client.newProcessInstanceQuery()
        .filter((f) -> f.variable((v) -> v.name("it").eq(5)))
        .sort((s) -> s.endDate().asc())
        .page((p) -> p.size(20).searchAfter(response.page().lastSortValues()))
        .send()
        .join();
```

romansmirnov commented:
```java
      final SearchQueryResponse<ProcessInstance> response = client.newProcessInstanceQuery()
          .sort((s) -> s.endDate().asc())
          .page((p) -> p.size(20))
          .send()
          .join();

      response.items().forEach(System.out::println);
```

returns

```
class ProcessInstance {
    tenantId: <default>
    key: 2251799813685252
    processVersion: 1
    bpmnProcessId: foo
    parentKey: null
    parentFlowNodeInstanceKey: null
    startDate: 2024-06-19T14:35:51.393+0000
    endDate: null
}
class ProcessInstance {
    tenantId: <default>
    key: 2251799813685260
    processVersion: 1
    bpmnProcessId: foo
    parentKey: null
    parentFlowNodeInstanceKey: null
    startDate: 2024-06-19T14:35:51.543+0000
    endDate: null
}
class ProcessInstance {
    tenantId: <default>
    key: 2251799813685268
    processVersion: 1
    bpmnProcessId: foo
    parentKey: null
    parentFlowNodeInstanceKey: null
    startDate: 2024-06-19T14:35:51.591+0000
    endDate: null
}
...
```

nathansandi commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/search/ProcessInstanceSortImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+  @Override
+  public ProcessInstanceSort startDate() {
+    return field("startDate");
```
[❓If I am not wrong, on Camunda Services, we refer those fields to sort, why we need re-do it here on the `SortImpl`?

romansmirnov replied:
Camunda Services is a server-side implementation, and the sorting there defines on which (index) field sorting must happen.

This code here is used on the client side, which eventually calls the `/process-instances/search` endpoint. At this point, the user constructs the search query being serialized into a JSON, for example:
```json
{
  "filter": {
    "processInstanceKey": 22456786958
  },
  "sort": [
    { "startDate": "ASC" }
  ],
  "page": {
    "limit": 10
  }
}
```
When the Process Instance REST controller receives that request, it must somehow map the client's search request to the actual service, i.e., based on the passed sorting field it will call the respective sort method in the service.

tmetzke commented on clients/java/src/main/java/io/camunda/zeebe/client/api/search/ProcessInstanceFilter.java:
```diff
@@ -0,0 +1,35 @@
+public interface ProcessInstanceFilter extends SearchRequestFilter {
+
+  /** Filter by process instance keys. */
+  ProcessInstanceFilter processInstanceKeys(final Long value, final Long... values);
```
❓ What is the benefit compared to just doing `ProcessInstanceFilter processInstanceKeys(final Long... values);`?

romansmirnov replied:
Ah yeah, I somehow took over that style from the ES/OS clients. I can simplify it.

tmetzke commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/search/ProcessInstanceFilterImpl.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Override
+  public ProcessInstanceFilter variable(final VariableValueFilter value) {
+    final VariableValueFilterImpl variableFilter = (VariableValueFilterImpl) value;
```
❓ Do you see a way to avoid those class cast instructions? We have many of them in this PR and I was wondering if we could avoid them. Code style-wise, this is not ideal. There could be a base interface that contains such a method that all the filter, sort, and page interfaces extend.

romansmirnov replied:
I tried a couple of variations and tried not to expose too much to the public API. Basically, I took the same approach of having a base interface that all classes implement and that provides that necessary method. However, this method would then be available to users as well, which I wanted to avoid. If you think this is not a problem, then okay, I can change that.

tmetzke replied:
Yeah, I was also wondering about whether we should expose this or not. I don't have a good answer right now from the top of my head. Exposing it publicly may be even worse 🤔

romansmirnov replied:
I think, I have an alternative to centralize that. I will finish it quickly; maybe it will make it a bit more handy.

romansmirnov replied:
Hm... what do you think? 9af53acba28f5dce1f4640f94ea387615e5c16ad At least, it makes it not worse than before.

tmetzke replied:
That's a lot better really since it actually checks the base type where this is implemented or extended from 👍

tmetzke commented on clients/java/src/main/java/io/camunda/zeebe/client/api/search/SearchRequestPage.java:
```diff
@@ -0,0 +1,33 @@
+
+public interface SearchRequestPage {
+
+  /** Start the page from. */
+  SearchRequestPage from(final Integer value);
```
❓ We don't have `from` defined in the REST API guidelines so far. What is this used for?

romansmirnov replied:
It can be used for a classic pagination without search after, however, I can remove it if not needed.

tmetzke replied:
Understood. Do we offer this on the controller/server side?

romansmirnov replied:
Yes, it offers it already.

tmetzke replied:
Ah OK, I didn't know 👍 Let's keep it then and add it to the REST guidelines 🙂

tmetzke replied:
I didn't add it to the guidelines yet because we weren't sure how complex it would be to add it to the search controllers.

tmetzke replied:
@romansmirnov, would you mind adding this to the REST guidelines by creating a proposal, describing how to define and use this in search and response? 🙂 I can take care of documenting this in the guidelines then 👍

tmetzke commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/ProcessInstanceController.java:
```diff
@@ -20,8 +29,31 @@ public class ProcessInstanceController {
+      final ProcessInstanceQuery query) {
+    try {
+      final var tenantIds = TenantAttributeHolder.tenantIds();
+      final var result =
+          processInstanceServices.withAuthentication((a) -> a.tenants(tenantIds)).search(query);
+      return ResponseEntity.ok(
+          SearchQueryResponseMapper.toProcessInstanceSearchQueryResponse(result).get());
```
❓ Just out of curiosity: would async searching be possible? Currently, this is a synchronous blocking call in the service's API. Have we discussed this anywhere in the past?

**Note**: this is not blocking the merge at all.

romansmirnov replied:
At least, I am not aware of whether this has been discussed or if any decision is being made. Technically, yes, we could also implement them asynchronously, ES and OS supports this as well.

tmetzke replied:
OK, let's defer that. Could be an interesting adjustment though to avoid blocking search requests and free I/O resources for search requests over REST.

tmetzke reviewed (approved):
Looks really good overall 👍 
I added a couple of understanding questions, nothing that really blocks the PR though.
